libdnf/module/ModulePackage.cpp

```diff
@@ -0,0 +1,358 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <iostream>
+#include <sstream>
+#include <utility>
+extern "C" {
+#include <solv/solver.h>
+}
+
+#include "ModulePackage.hpp"
+#include "libdnf/utils/File.hpp"
+
+static void setSovable(Pool *pool, Solvable *solvable, std::string name, std::string stream, std::string version, std::string context, const char * arch)
+{
+    std::ostringstream ss;
+    // create solvable with:
+    //   Name: $name:$stream:$version:$context
+    //   Version: 0
+    //   Arch: $arch
+    ss << name << ":" << stream << ":" << version << ":" << context;
+    solvable_set_str(solvable, SOLVABLE_NAME, ss.str().c_str());
+    solvable_set_str(solvable, SOLVABLE_EVR, "0");
+    // TODO Test can be remove when modules will be always with arch
+    solvable_set_str(solvable, SOLVABLE_ARCH, arch ? arch : "noarch");
+    // create Provide: module($name)
+    ss.str(std::string());
+    ss << "module(" << name << ")";
+    auto depId = pool_str2id(pool, ss.str().c_str(), 1);
+    solvable_add_deparray(solvable, SOLVABLE_PROVIDES, depId, -1);
+
+    // create Provide: module($name:$stream)
+    ss.str(std::string());
+    ss << "module(" << name << ":" << stream << ")";
+    depId = pool_str2id(pool, ss.str().c_str(), 1);
+    solvable_add_deparray(solvable, SOLVABLE_PROVIDES, depId, -1);
+
+    // create Provide: module($name:$stream:$version)
+    ss.str(std::string());
+    ss << "module(" << name << ":" << stream << ":" << version << ")";
+    depId = pool_str2id(pool, ss.str().c_str(), 1);
+    solvable_add_deparray(solvable, SOLVABLE_PROVIDES, depId, -1);
+}
+
+ModulePackage::ModulePackage(Pool *pool, Repo *repo, const std::shared_ptr<ModuleMetadata> &metadata)
+        : metadata(metadata)
+        , state(ModuleState::UNKNOWN)
+        , pool(pool)
+{
+    id = repo_add_solvable(repo);
+    Solvable *solvable = pool_id2solvable(pool, id);
+
+    setSovable(pool, solvable, getName(), getStream(), getVersion(), getContext(), metadata->getArchitecture());
+    createDependencies(solvable);
+}
+
+/**
+ * @brief Create stream dependencies based on modulemd metadata.
+ */
+void ModulePackage::createDependencies(Solvable *solvable) const
+{
+    std::ostringstream ss;
+    Id depId;
+
+    for (const auto &dependency : getModuleDependencies()) {
```

Comment:
The function getModuleDependencies() should return vector with already formated dependencies (string/ char*).